File: Inxi.NET/Core/InxiHardwareInfo.vb

'    Inxi.NET  Copyright (C) 2020-2021  EoflaOE
'
'    This file is part of Inxi.NET
'
'    Inxi.NET is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Inxi.NET is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Management
Imports Claunia.PropertyList
Imports Newtonsoft.Json.Linq

Public Class HardwareInfo

    ''' <summary>
    ''' List of hard drives detected
    ''' </summary>
    Public ReadOnly HDD As New Dictionary(Of String, HardDrive)
    ''' <summary>
    ''' List of logical hard drive partitions detected
    ''' </summary>
    Public ReadOnly LogicalParts As New Dictionary(Of String, WindowsLogicalPartition)
    ''' <summary>
    ''' List of processors detected
    ''' </summary>
    Public ReadOnly CPU As New Dictionary(Of String, Processor)
    ''' <summary>
    ''' List of graphics cards detected
    ''' </summary>
    Public ReadOnly GPU As New Dictionary(Of String, Graphics)
    ''' <summary>
    ''' List of sound cards detected
    ''' </summary>
    Public ReadOnly Sound As New Dictionary(Of String, Sound)
    ''' <summary>
    ''' List of network cards detected
    ''' </summary>
    Public ReadOnly Network As New Dictionary(Of String, Network)
    ''' <summary>
    ''' System information
    ''' </summary>
    Public ReadOnly System As SystemInfo
    ''' <summary>
    ''' Machine information
    ''' </summary>
    Public ReadOnly Machine As MachineInfo
    ''' <summary>
    ''' BIOS information
    ''' </summary>
    Public ReadOnly BIOS As BIOS
    ''' <summary>
    ''' RAM information
    ''' </summary>
    Public ReadOnly RAM As PCMemory

    ''' <summary>
    ''' Inxi token used for hardware probe
    ''' </summary>
    Friend InxiToken As JToken
    ''' <summary>
    ''' SystemProfiler token used for hardware probe
    ''' </summary>
    Friend SystemProfilerToken As NSArray

    ''' <summary>
    ''' Initializes a new instance of hardware info
    ''' </summary>
    Sub New(InxiPath As String, ParseFlags As InxiHardwareType)
        If IsUnix() Then
            If IsMacOS() Then
                'Start the SystemProfiler process
                Dim SystemProfilerProcess As New Process
                Dim SystemProfilerProcessInfo As New ProcessStartInfo With {.FileName = "/usr/sbin/system_profiler", .Arguments = "SPSoftwareDataType SPAudioDataType SPHardwareDataType SPNetworkDataType SPStorageDataType SPDisplaysDataType -xml",
                                                                            .CreateNoWindow = True,
                                                                            .UseShellExecute = False,
                                                                            .WindowStyle = ProcessWindowStyle.Hidden,
                                                                            .RedirectStandardOutput = True}
                SystemProfilerProcess.StartInfo = SystemProfilerProcessInfo
                Debug("Starting system_profiler with ""SPSoftwareDataType SPAudioDataType SPHardwareDataType SPNetworkDataType SPStorageDataType SPDisplaysDataType -xml""...")
                SystemProfilerProcess.Start()
                SystemProfilerProcess.WaitForExit(10000)
                SystemProfilerToken = PropertyListParser.Parse(Text.Encoding.Default.GetBytes(SystemProfilerProcess.StandardOutput.ReadToEnd))
                Debug("Token parsed.")
            Else
                'Start the Inxi process
                Dim InxiProcess As New Process
                Dim InxiProcessInfo As New ProcessStartInfo With {.FileName = InxiPath, .Arguments = "-Fxx --output json --output-file print",
                                                                  .CreateNoWindow = True,
                                                                  .UseShellExecute = False,
                                                                  .WindowStyle = ProcessWindowStyle.Hidden,
                                                                  .RedirectStandardOutput = True}
                InxiProcess.StartInfo = InxiProcessInfo
                Debug("Starting inxi with ""-Fxx --output json --output-file print""...")
                InxiProcess.Start()
                InxiProcess.WaitForExit()
                InxiToken = JToken.Parse(InxiProcess.StandardOutput.ReadToEnd)
                Debug("Token parsed.")
            End If
        End If

        'Ready variables
        Dim HDDParsed As New Dictionary(Of String, HardwareBase)
        Dim Logicals As New Dictionary(Of String, WindowsLogicalPartition)
        Dim CPUParsed As New Dictionary(Of String, HardwareBase)
        Dim GPUParsed As New Dictionary(Of String, HardwareBase)
        Dim SoundParsed As New Dictionary(Of String, HardwareBase)
        Dim NetParsed As New Dictionary(Of String, HardwareBase)
        Dim RAMParsed As PCMemory
        Dim BIOSParsed As BIOS
        Dim SystemParsed As SystemInfo
        Dim MachineParsed As MachineInfo

        'Parse hardware starting from HDD
        If ParseFlags.HasFlag(InxiHardwareType.HardDrive) Then
            Debug("Parsing HDD...")
            Dim BaseParser As New HardDriveParser
            HDDParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.HardDrive)
        End If

        'Logical partitions
        If Not IsUnix() And ParseFlags.HasFlag(InxiHardwareType.HardDriveLogical) Then
            Debug("Parsing logical partitions...")
            Logicals = ParsePartitions(New ManagementObjectSearcher("SELECT * FROM Win32_LogicalDisk"))
            RaiseParsedEvent(InxiHardwareType.HardDriveLogical)
        End If

        'Processor
        If ParseFlags.HasFlag(InxiHardwareType.Processor) Then
            Debug("Parsing CPU...")
            Dim BaseParser As New ProcessorParser
            CPUParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.Processor)
        End If

        'Graphics
        If ParseFlags.HasFlag(InxiHardwareType.Graphics) Then
            Debug("Parsing GPU...")
            Dim BaseParser As New GraphicsParser
            GPUParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.Graphics)
        End If

        'Sound
        If ParseFlags.HasFlag(InxiHardwareType.Sound) Then
            Debug("Parsing sound...")
            Dim BaseParser As New SoundParser
            SoundParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.Sound)
        End If

        'Network
        If ParseFlags.HasFlag(InxiHardwareType.Network) Then
            Debug("Parsing network...")
            Dim BaseParser As New NetworkParser
            NetParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.Network)
        End If

        'PC Memory
        If ParseFlags.HasFlag(InxiHardwareType.PCMemory) Then
            Debug("Parsing RAM...")
            Dim BaseParser As New PCMemoryParser
            RAMParsed = BaseParser.Parse(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.PCMemory)
        End If

        'BIOS
        If ParseFlags.HasFlag(InxiHardwareType.BIOS) Then
            Debug("Parsing BIOS...")
            Dim BaseParser As New BIOSParser
            BIOSParsed = BaseParser.Parse(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.BIOS)
        End If

        'System
        If ParseFlags.HasFlag(InxiHardwareType.System) Then
            Debug("Parsing system...")
            Dim BaseParser As New SystemParser
            SystemParsed = BaseParser.Parse(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.System)
        End If

        'Machine
        If ParseFlags.HasFlag(InxiHardwareType.Machine) Then
            Debug("Parsing machine...")
            Dim BaseParser As New MachineParser
            MachineParsed = BaseParser.Parse(InxiToken, SystemProfilerToken)
            RaiseParsedEvent(InxiHardwareType.Machine)
        End If

        'Add the base to the correct type
        Dim HDDProcessed As New Dictionary(Of String, HardDrive)
        Dim CPUProcessed As New Dictionary(Of String, Processor)
        Dim GPUProcessed As New Dictionary(Of String, Graphics)
        Dim SoundProcessed As New Dictionary(Of String, Sound)
        Dim NetProcessed As New Dictionary(Of String, Network)

        'Hard drive
        For Each Parsed As String In HDDParsed.Keys
            Dim FinalHardware As HardDrive = HDDParsed(Parsed)
            HDDProcessed.Add(Parsed, FinalHardware)
        Next

        'Processor
        For Each Parsed As String In CPUParsed.Keys
            Dim FinalHardware As Processor = CPUParsed(Parsed)
            CPUProcessed.Add(Parsed, FinalHardware)
        Next

        'Graphics
        For Each Parsed As String In GPUParsed.Keys
            Dim FinalHardware As Graphics = GPUParsed(Parsed)
            GPUProcessed.Add(Parsed, FinalHardware)
        Next

        'Sound
        For Each Parsed As String In SoundParsed.Keys
            Dim FinalHardware As Sound = SoundParsed(Parsed)
            SoundProcessed.Add(Parsed, FinalHardware)
        Next

        'Network
        For Each Parsed As String In NetParsed.Keys
            Dim FinalHardware As Network = NetParsed(Parsed)
            NetProcessed.Add(Parsed, FinalHardware)
        Next

        'Install parsed information to current instance
        HDD = HDDProcessed
        If Not IsUnix() Then LogicalParts = Logicals
        CPU = CPUProcessed
        GPU = GPUProcessed
        Sound = SoundProcessed
        Network = NetProcessed
#Disable Warning BC42104
        RAM = RAMParsed
        BIOS = BIOSParsed
        System = SystemParsed
        Machine = MachineParsed
#Enable Warning BC42104
        Debug("Parsed information installed.")
    End Sub

End Class

File: Inxi.NET/Core/InxiHardwareInfo.cs
using Claunia.PropertyList;
using Extensification.StringExts;
using InxiFrontend.Base;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Management;
using System.Text;

namespace InxiFrontend
{
    /// <summary>
    /// Hardware information class
    /// </summary>
    public class HardwareInfo
    {
        [JsonProperty()]
        /// <summary>
        /// List of hard drives detected
        /// </summary>
        public ConcurrentDictionary<string, HardDrive> HDD { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// List of logical hard drive partitions detected
        /// </summary>
        public ConcurrentDictionary<string, WindowsLogicalPartition> LogicalParts { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// List of processors detected
        /// </summary>
        public ConcurrentDictionary<string, Processor> CPU { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// List of graphics cards detected
        /// </summary>
        public ConcurrentDictionary<string, Graphics> GPU { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// List of sound cards detected
        /// </summary>
        public ConcurrentDictionary<string, Sound> Sound { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// List of network cards detected
        /// </summary>
        public ConcurrentDictionary<string, Network> Network { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// List of batteries detected
        /// </summary>
        public HashSet<Battery> Battery { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// System information
        /// </summary>
        public SystemInfo System { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// Machine information
        /// </summary>
        public MachineInfo Machine { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// BIOS information
        /// </summary>
        public BIOS BIOS { get; private set; }
        [JsonProperty()]
        /// <summary>
        /// RAM information
        /// </summary>
        public PCMemory RAM { get; private set; }
        [JsonIgnore()]
        /// <summary>
        /// Inxi token used for hardware probe
        /// </summary>
        internal JToken InxiToken;
        [JsonIgnore()]
        /// <summary>
        /// SystemProfiler token used for hardware probe
        /// </summary>
        internal NSArray SystemProfilerToken;

        /// <summary>
        /// Initializes a new instance of hardware info
        /// </summary>
        public HardwareInfo(string InxiPath, InxiHardwareType ParseFlags)
        {
            if (InxiInternalUtils.IsUnix())
            {
                if (InxiInternalUtils.IsMacOS())
                {
                    // Start the SystemProfiler process
                    var SystemProfilerProcess = new Process();
                    var SystemProfilerProcessInfo = new ProcessStartInfo()
                    {
                        FileName = "/usr/sbin/system_profiler",
                        Arguments = "SPSoftwareDataType SPAudioDataType SPHardwareDataType SPNetworkDataType SPStorageDataType SPDisplaysDataType -xml",
                        CreateNoWindow = true,
                        UseShellExecute = false,
                        WindowStyle = ProcessWindowStyle.Hidden,
                        RedirectStandardOutput = true
                    };
                    SystemProfilerProcess.StartInfo = SystemProfilerProcessInfo;
                    InxiTrace.Debug("Starting system_profiler with \"SPSoftwareDataType SPAudioDataType SPHardwareDataType SPNetworkDataType SPStorageDataType SPDisplaysDataType -xml\"...");
                    SystemProfilerProcess.Start();
                    SystemProfilerProcess.WaitForExit(10000);
                    SystemProfilerToken = (NSArray)PropertyListParser.Parse(Encoding.Default.GetBytes(SystemProfilerProcess.StandardOutput.ReadToEnd()));
                    InxiTrace.Debug("Token parsed.");
                }
                else
                {
                    // Start the Inxi process
                    var InxiProcess = new Process();
                    var InxiProcessInfo = new ProcessStartInfo()
                    {
                        FileName = InxiPath,
                        Arguments = "-Fxx --output json --output-file print",
                        CreateNoWindow = true,
                        UseShellExecute = false,
                        WindowStyle = ProcessWindowStyle.Hidden,
                        RedirectStandardOutput = true
                    };
                    InxiProcess.StartInfo = InxiProcessInfo;
                    InxiTrace.Debug("Starting inxi with \"-Fxx --output json --output-file print\"...");
                    InxiProcess.Start();
                    InxiProcess.WaitForExit();
                    var lines = InxiProcess.StandardOutput.ReadToEnd().SplitNewLines();
                    InxiToken = JToken.Parse(lines[lines.Length - 1]);
                    InxiTrace.Debug("Token parsed.");
                }
            }

            // Ready variables
            var HDDParsed = new Dictionary<string, IHardware>();
            var Logicals = new Dictionary<string, WindowsLogicalPartition>();
            var CPUParsed = new Dictionary<string, IHardware>();
            var GPUParsed = new Dictionary<string, IHardware>();
            var SoundParsed = new Dictionary<string, IHardware>();
            var NetParsed = new Dictionary<string, IHardware>();
            var BatteryParsed = new List<IHardware>();
            var RAMParsed = default(PCMemory);
            var BIOSParsed = default(BIOS);
            var SystemParsed = default(SystemInfo);
            var MachineParsed = default(MachineInfo);

            // Parse hardware starting from HDD
            if (ParseFlags.HasFlag(InxiHardwareType.HardDrive))
            {
                InxiTrace.Debug("Parsing HDD...");
                var BaseParser = new HardDriveParser();
                HDDParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.HardDrive);
            }

            // Logical partitions
            if (!InxiInternalUtils.IsUnix() && ParseFlags.HasFlag(InxiHardwareType.HardDriveLogical))
            {
                InxiTrace.Debug("Parsing logical partitions...");
                Logicals = WindowsLogicalPartitionParser.ParsePartitions(new ManagementObjectSearcher("SELECT * FROM Win32_LogicalDisk"));
                InxiTrace.RaiseParsedEvent(InxiHardwareType.HardDriveLogical);
            }

            // Processor
            if (ParseFlags.HasFlag(InxiHardwareType.Processor))
            {
                InxiTrace.Debug("Parsing CPU...");
                var BaseParser = new ProcessorParser();
                CPUParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.Processor);
            }

            // Graphics
            if (ParseFlags.HasFlag(InxiHardwareType.Graphics))
            {
                InxiTrace.Debug("Parsing GPU...");
                var BaseParser = new GraphicsParser();
                GPUParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.Graphics);
            }

            // Sound
            if (ParseFlags.HasFlag(InxiHardwareType.Sound))
            {
                InxiTrace.Debug("Parsing sound...");
                var BaseParser = new SoundParser();
                SoundParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.Sound);
            }

            // Network
            if (ParseFlags.HasFlag(InxiHardwareType.Network))
            {
                InxiTrace.Debug("Parsing network...");
                var BaseParser = new NetworkParser();
                NetParsed = BaseParser.ParseAll(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.Network);
            }

            // PC Memory
            if (ParseFlags.HasFlag(InxiHardwareType.PCMemory))
            {
                InxiTrace.Debug("Parsing RAM...");
                var BaseParser = new PCMemoryParser();
                RAMParsed = (PCMemory)BaseParser.Parse(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.PCMemory);
            }

            // BIOS
            if (ParseFlags.HasFlag(InxiHardwareType.BIOS))
            {
                InxiTrace.Debug("Parsing BIOS...");
                var BaseParser = new BIOSParser();
                BIOSParsed = (BIOS)BaseParser.Parse(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.BIOS);
            }

            // System
            if (ParseFlags.HasFlag(InxiHardwareType.System))
            {
                InxiTrace.Debug("Parsing system...");
                var BaseParser = new SystemParser();
                SystemParsed = (SystemInfo)BaseParser.Parse(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.System);
            }

            // Machine
            if (ParseFlags.HasFlag(InxiHardwareType.Machine))
            {
                InxiTrace.Debug("Parsing machine...");
                var BaseParser = new MachineParser();
                MachineParsed = (MachineInfo)BaseParser.Parse(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.Machine);
            }

            // Battery
            if (ParseFlags.HasFlag(InxiHardwareType.Battery))
            {
                InxiTrace.Debug("Parsing battery...");
                var BaseParser = new BatteryParser();
                BatteryParsed = BaseParser.ParseAllToList(InxiToken, SystemProfilerToken);
                InxiTrace.RaiseParsedEvent(InxiHardwareType.Battery);
            }

            // Add the base to the correct type
            var HDDProcessed = new Dictionary<string, HardDrive>();
            var CPUProcessed = new Dictionary<string, Processor>();
            var GPUProcessed = new Dictionary<string, Graphics>();
            var SoundProcessed = new Dictionary<string, Sound>();
            var NetProcessed = new Dictionary<string, Network>();
            var BatteryProcessed = new List<Battery>();

            // Hard drive
            foreach (string Parsed in HDDParsed.Keys)
            {
                HardDrive FinalHardware = (HardDrive)HDDParsed[Parsed];
                HDDProcessed.Add(Parsed, FinalHardware);
            }

            // Processor
            foreach (string Parsed in CPUParsed.Keys)
            {
                Processor FinalHardware = (Processor)CPUParsed[Parsed];
                CPUProcessed.Add(Parsed, FinalHardware);
            }

            // Graphics
            foreach (string Parsed in GPUParsed.Keys)
            {
                Graphics FinalHardware = (Graphics)GPUParsed[Parsed];
                GPUProcessed.Add(Parsed, FinalHardware);
            }

            // Sound
            foreach (string Parsed in SoundParsed.Keys)
            {
                Sound FinalHardware = (Sound)SoundParsed[Parsed];
                SoundProcessed.Add(Parsed, FinalHardware);
            }

            // Network
            foreach (string Parsed in NetParsed.Keys)
            {
                Network FinalHardware = (Network)NetParsed[Parsed];
                NetProcessed.Add(Parsed, FinalHardware);
            }

            // Battery
            foreach (Battery Parsed in BatteryParsed.Cast<Battery>())
                BatteryProcessed.Add(Parsed);

            // Install parsed information to current instance
            HDD = new(HDDProcessed);
            if (!InxiInternalUtils.IsUnix())
                LogicalParts = new(Logicals);
            CPU = new(CPUProcessed);
            GPU = new(GPUProcessed);
            Sound = new(SoundProcessed);
            Network = new(NetProcessed);
            Battery = new(BatteryProcessed);
            RAM = RAMParsed;
            BIOS = BIOSParsed;
            System = SystemParsed;
            Machine = MachineParsed;
            InxiTrace.Debug("Parsed information installed.");
        }

        [JsonConstructor()]
        public HardwareInfo()
        {
        }
    }
}